File: make-the-string-great/solution1544.go
// https://leetcode.com/problems/make-the-string-great/

package main

import (
	"fmt"
	"unicode"
)

// T: O(n), S: O(n)
func makeGood(s string) string {
	stack := make([]rune, 0)
	for _, ch := range s {
		if len(stack) != 0 && areSameCharsDiffCases(ch, stack[len(stack)-1]) {
			stack = stack[:len(stack)-1]
		} else {
			stack = append(stack, ch)
		}
	}
	return string(stack)
}

func areSameCharsDiffCases(ch1, ch2 rune) bool {
	if ch1 == ch2 {
		return false
	}
	if unicode.ToLower(ch1) != unicode.ToLower(ch2) {
		return false
	}
	return true
}

func main() {
	s1 := "leEeetcode"
	s2 := "abBAcC"
	s3 := "s"
	fmt.Println(makeGood(s1)) // "leetcode"
	fmt.Println(makeGood(s2)) // ""
	fmt.Println(makeGood(s3)) // "s"
}

File: make-the-string-great/Solution1544.java
// https://leetcode.com/problems/make-the-string-great/

import java.util.Stack;

class Solution1544 {

    // T: O(n), S: O(n)
    public String makeGood(String str) {
        var stack = new Stack<Character>();
        for (int i = 0; i < str.length(); i++) {
            char ch = str.charAt(i);
            if (!stack.isEmpty() && areSameCharsDiffCases(ch, stack.peek())) {
                stack.pop();
            } else {
                stack.push(ch);
            }
        }
        var sb = new StringBuilder();
        for (var ch : stack) {
            sb.append(ch);
        }
        return sb.toString();
    }

    private boolean areSameCharsDiffCases(char ch1, char ch2) {
        if (ch1 == ch2) {
            return false;
        }
        if (Character.toLowerCase(ch1) != Character.toLowerCase(ch2)) {
            return false;
        }
        return true;
    }

    public static void main(String[] args) {
        var sol = new Solution1544();
        var s1 = "leEeetcode";
        var s2 = "abBAcC";
        var s3 = "s";
        System.out.println(sol.makeGood(s1)); // "leetcode"
        System.out.println(sol.makeGood(s2)); // ""
        System.out.println(sol.makeGood(s3)); // "s"
    }

}
